File: app/src/main/java/com/bogdan/codeforceswatcher/features/actions/models/ActionItem.kt
package com.bogdan.codeforceswatcher.features.actions.models

import android.text.SpannableStringBuilder
import com.bogdan.codeforceswatcher.CwApp
import com.bogdan.codeforceswatcher.R
import com.bogdan.codeforceswatcher.features.users.colorTextByUserRank
import com.bogdan.codeforceswatcher.util.convertFromHtml
import io.xorum.codeforceswatcher.features.actions.models.CFAction
import io.xorum.codeforceswatcher.network.responses.PinnedPost
import io.xorum.codeforceswatcher.util.FeedbackData
import io.xorum.codeforceswatcher.util.avatar

sealed class ActionItem {

    class CommentItem(action: CFAction) : ActionItem() {

        var commentatorHandle: CharSequence
        var title: String = action.blogEntry.title.convertFromHtml()
        var content: String
        var commentatorAvatar: String
        val time: Long = action.timeSeconds
        val link = action.link

        init {
            val comment = action.comment ?: throw NullPointerException()
            commentatorAvatar = avatar(comment.commentatorAvatar ?: throw IllegalStateException())
            commentatorHandle = buildHandle(comment.commentatorHandle, comment.commentatorRank)
            content = comment.text.convertFromHtml()
        }

        private fun buildHandle(handle: String, rank: String?): CharSequence {
            val colorHandle = colorTextByUserRank(handle, rank)
            val commentedByString = CwApp.app.getString(R.string.commented_by)
            val handlePosition = commentedByString.indexOf("%1\$s")

            return SpannableStringBuilder(commentedByString)
                    .replace(handlePosition, handlePosition + "%1\$s".length, colorHandle)
        }
    }

    class BlogEntryItem(action: CFAction) : ActionItem() {

        val authorHandle: CharSequence
        val blogTitle: String
        val authorAvatar: String
        val time: Long = action.timeSeconds
        val link = action.link

        init {
            with(action) {
                authorAvatar = avatar(blogEntry.authorAvatar ?: throw IllegalStateException())
                authorHandle = colorTextByUserRank(blogEntry.authorHandle, blogEntry.authorRank)
                blogTitle = blogEntry.title.convertFromHtml()
            }
        }
    }

    class PinnedItem(pinnedPost: PinnedPost) : ActionItem() {
        val title = pinnedPost.title
        val link = pinnedPost.link
    }

    class FeedbackItem(feedbackData: FeedbackData): ActionItem() {
        val textPositiveButton = feedbackData.textPositiveButton
        val textNegativeButton = feedbackData.textNegativeButton
        val textTitle = feedbackData.textTitle
        val positiveButtonClick = feedbackData.positiveButtonClick
        val negativeButtonClick = feedbackData.negativeButtonClick
        val neutralButtonClick = feedbackData.neutralButtonClick
    }

    object Stub : ActionItem()
}

File: ios/Codeforces Watcher/Application/features/actions/models/ActionItem.swift
//
//  ActionItem.swift
//  Codeforces Watcher
//
//  Created by Den Matyash on 1/8/20.
//  Copyright © 2020 xorum.io. All rights reserved.
//

import Foundation
import HTMLString
/*
enum ActionItem {
    struct CommentItem {
        var shareText: String
        var link: String
        var commentatorHandle: NSMutableAttributedString
        var blogTitle: String
        var content: String
        var time: String
        var commentatorAvatar: String
        
        init(action: CFAction, rank: String?, avatar: String) {
            self.blogTitle = action.blogEntry.title!.beautify()
            self.time = TimeInterval((Int(Date().timeIntervalSince1970) - Int(action.timeSeconds!))).socialDate ?? ""
            self.content = action.comment!.text!.beautify()
            self.commentatorHandle = colorTextByUserRank(text: action.comment!.commentatorHandle!, rank: rank)
            self.commentatorAvatar = "https:" + avatar
            self.link = "\(codeforcesLink)blog/entry/\(action.blogEntry.id!)?#comment-\(action.comment!.id!)"
            self.shareText = """
            \(blogTitle) - \(link)
            
            Shared through Codeforces Watcher. Find it on App Store.
            """
        }
    }

    struct BlogEntryItem {
        var shareText: String
        var link: String
        var authorHandle: NSMutableAttributedString
        var blogTitle: String
        var time: String
        var authorAvatar: String
        
        init(action: CFAction, rank: String?, avatar: String) {
            self.blogTitle = action.blogEntry.title!.beautify()
            self.time = TimeInterval((Int(Date().timeIntervalSince1970) - Int(action.timeSeconds!))).socialDate ?? ""
            self.authorHandle = colorTextByUserRank(text: action.blogEntry!.authorHandle!, rank: rank)
            self.authorAvatar = "https:" + avatar
            self.link = "\(codeforcesLink)blog/entry/\(action.blogEntry.id!)"
            self.shareText = """
            \(blogTitle) - \(link)
            
            Shared through Codeforces Watcher. Find it on App Store.
            """
        }
    }
    
    case comment(CommentItem)
    case blogEntry(BlogEntryItem)
}
*/
